## Case studies and analyses {#section:casestudies}

During development, the still incomplete pipeline was applied to several
datasets as a means of exploration. This has helped to find bugs,
usability problems, discover other applications and guide the UMGAP to
its final form. This section describes the analysis of these datasets,
as it would be done using the released version of the UMGAP, not as it
was executed originally. As such, this section may be used as a guide to
compose a new pipeline for a similar analysis.

### The preconfigured pipelines {#section:preconf}

The primary and easiest procedure to run the UMGAP on your data, and the
last addition to the pipeline, is by using one of the six preconfigured
pipelines. They are six diverging ways of configuring the pipeline,
picked to cover the most probable use cases. Should your data fall within
such a use case, running the pipeline is reduced to running two or three
scripts.

The first is a setup script called `umgap-setup.sh`. It is an
interactive script which helps you get the required databases and
optional external tools in the right place. In general, the setup script
only needs to run once, but it can also be used to verify the file
locations and download newer versions of the data.

Following code snippet shows the interaction with the setup script.
It starts out by asking the relevant questions, without delay. At the
end, it downloads the relevant files in sequence, without further
interaction. In this snippet, we choose to download the taxonomy and the
tryptic index, as this example will use a tryptic pipeline.

```shell
$ umgap-setup.sh -f /opt/FragGeneScanPlusPlus
Use '/home/user/.config/unipept' as configuration directory? [y]/n y
Created directory /home/user/.config/unipept.
Found, tested and remembered the FragGeneScan++ location.
Use '/home/user/.local/share/unipept' as data directory? [y]/n y
Created directory /home/user/.local/share/unipept.
Checking the latest version on the server.
Latest version is 2020-12-02.

For any type of analysis, you need a taxonomony. For mapping tryptic or
9-mer peptides to it, you need the respective index file of the same
version.

Would you like to download the taxonomy from 2020-12-02 (115MiB)? [y]/n y
Would you like to download the tryptic index from 2020-12-02 (12GiB)? [y]/n y
Would you like to download the 9-mer index from 2020-12-02 (152GiB)? [y]/n n
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  114M  100  114M    0     0  6477k      0  0:00:18  0:00:18 --:--:-- 6332k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   12G  100   12G    0     0  5076k      0  0:34:12  0:34:12 --:--:-- 5790k
```

The script also allows further configuration via some flags.

`-c dir`
  ~ Set a different configuration directory.

`-f dir`
  ~ Link the location of your FragGeneScan installation.

`-d dir`
  ~ Set a different location for downloading the database files. This
    can be on a separate disk.

`-y`
  ~ Download all files of the latest version without asking for
    confirmation.

This last flag, `-y`, is especially useful when running the setup script
from another script, e.g. an automatic update script.

The second script is the main analysis script, aptly called
`umgap-analyse.sh`. It will run a selected pipeline on the input data
using the databases downloaded by the setup script. Consider for
instance a 100 paired-end reads sampled from a dataset generated by
@lindgreen for their MetaBenchmark. With these paired-end reads in the
`A1.fq` and `A2.fq` files, the commands are as follows.

```shell
$ head -4 A1.fq
@1198114###CP002480-_Acidobacteria_733918/1
ATCGCGCACGCGGCCGATGCCCCAGAAGAGATTGACAGCGGTGGGGCGGGCGGCGGCGAGGTGGTCGCAGATCTCGGCGACCTCTGCGTTGAGGGTCGGG
+
AADEGGAGIIEIHFHKHJKKJKHIJJIIKCJAHBFKKFHIJIF;JIH5DE$I>CBE$E:FGEJCGABEEECCDKD?D?C$ECEEE;CEEDEEAD=?ED$D
$ umgap-analyse.sh -1 A1.fq -2 A2.fq -t tryptic-precision -z -o tryptic-precision-output.fa.gz
$ zcat tryptic-precision-output.fa.gz | head -2
>1198114###CP002480-_Acidobacteria_733918/1_1_100_-
1
```

It seems the pipeline cannot be more specific than root (Taxon ID 1) for
the first read.

The analysis script was passed a single series of arguments here (`-1`
to through `-o`), but it accepts multiple such sequences (along with the
other flags specified below) to combine multiple analyses into a single
command.

`-c dir`
  ~ Set a different configuration directory.

`-1 file`
  ~ Single ended FASTA or first pair-ended FASTQ input file, optionally
    compressed.

`-2 file`
  ~ Second pair-ended FASTQ input file, optionally compressed.

`-t`
  ~ Type of the analysis. This flag allows you to select one of the
    preconfigured pipelines.

    - `high-precision` is the default pipeline and is optimized for high
      precision identifications on your metagenomics reads.

    - `max-precision` is optimized for highest precision, with a small
      setback on sensitivity.

    - `tryptic-precision` is made for fast analyses on your laptop.
      Fewer results, but accurate.

    - `high-sensitivity` is optimized for high sensitivity
      identifications on metagenomics reads.

    - `max-sensitivity` is optimized for highest sensitivity, with a
      small setback on precision.

    - `tryptic-sensitivity` is made for fast analyses on your laptop.
      Get a quick overview of your data.

`-z`
  ~ Request compression of the output file.

`-o file`
  ~ The output file.

The output file contains a single taxonomic prediction for every read
after the analysis script. To interpret these results without parsing
and scripting visualizations, the `umgap-visualize.sh` script is
provided. This visualization script can create importable CSV frequency
tables and interactive visualizations. The latter can be stored locally
or hosted online. The following snippet creates all three in turn.

```shell
$ umgap-visualize.sh -t -r phylum tryptic-precision-output.fa.gz
taxon id,taxon name,tryptic-precision-output.fa.gz
1117,Cyanobacteria,2
57723,Acidobacteria,4
201174,Actinobacteria,4
1224,Proteobacteria,5
1,root,160
$ umgap-visualize.sh -w tryptic-sensitivity-output.fa.gz > tryptic-sensitivity.html
$ umgap-visualize.sh -u tryptic-sensitivity-output.fa.gz
tryptic-sensitivity-output.fa.gz: https://bl.ocks.org/11b7809d6754b9530cf1a49d93a8d568
```

The CSV tables contain a record for each taxon found in the sample.
A record contains, in order, the taxon ID, (for convenience) the
taxon name and the number of reads assigned to this taxon or below.
All records will be at the same specified taxon rank or at root
(unidentified).

`-t`
  ~ Output a CSV frequency table on species rank.

`-w`
  ~ Output an HTML web page of an interactive visualization.

`-u`
  ~ Print a shareable URL to a online interactive visualization.

`-r rank`
  ~ Set the rank for the CSV frequency table (default: species).

### A home brew taxonomic analysis pipeline {#section:homebrew}

While the preconfigured pipelines make running the UMGAP relatively
easy, there are only six of them. Part of the reason for using a
pipeline is to allow composing many diverse metagenomic pipelines
with the same set of tools. This section builds up a custom pipeline,
demonstrating how all tools work together.

The 3 reads of the same dataset as in \ref{section:preconf} will
be used. This is a paired-end dataset, stored in two FASTQ files.
Since the pipeline operates on FASTA files, the first step is a format
conversion.

```shell
$ umgap fastq2fasta A1.fq A2.fq | tee preprocessed.fa
>read1/1
ATCGCGCACGCGGCCGATGCCCCAGAAGAGATTGACAGCGGTGGGGCGGGCGGCGGCGAGGTGGTCGCAGATCTCGGCGACCTCTGCGTTGAGGGTCGGG
>read1/2
AAGATGGCGACTGGATGATGATCGCGCGGCAGGCCACCATCTACGATCCCGCAGTGAAGCATTACTAACCATGATGCGCAACGACTCGCTGTCAGAGCTA
>read2/1
AGATTGCTGGTGCGGGTGCTCTGCCGGGCTTCTTCATCTCGGACCGCGCATCCGATGCGCACACGGCACAGTAGGTATGCGGTGAGAGCACCTCGCTTTT
>read2/2
CGCAATCTTGCGGCGCACCGCGATCAAATCGGGATAGTCCGGTGTGTAACGCGTGCTCAGGTCATCTGCCCGTACCCGCAATACATTCAACTGTGTTTTA
>read3/1
CCCTCAAGCCCGATCGAACTTTCCTACTGCCCGACCTCAGCGCGGAGCATTACCGAATCGTCGTCAACAATCTCCCCGATGGCTTCTATGTGAACTCCAT
>read3/2
CAAAATGGAGTGGGCGCTACTGCTCCGTGAGCAGGGTCAGTTGGAGGGATTGGGGCGTGCGCTCGGGGATGCTAATGGCGGTGCCCCTGGATTCCTGAGA
```

This command interleaves the given FASTQ files to output a single
FASTA stream, in which the paired ends alternate. This FASTA stream is
a stream of DNA reads. Each read needs to be translated to an amino
acid sequence. While any gene prediction tool could be used here, the
UMGAP was tested with FragGeneScan and its variants.

```shell
$ FGS -s preprocessed.fa -o predicted-genes -w0 -t illumina_10 -p 16 > /dev/null
$ rm predicted-genes.fnn predicted-genes.out # we don't use these files
```

Alternatively, the built-in translation tool can be used to translate
all six frames of each read. Whichever method used, it results in a file
containing protein fragments.

```shell
$ umgap translate -a < preprocessed.fa | tee predicted-genes.faa
>read1/1|1
IAHAADAPEEIDSGGAGGGEVVADLGDLCVEGR
>read1/1|2
SRTRPMPQKRLTAVGRAAARWSQISATSALRVG
>read1/1|3
RARGRCPRRD*QRWGGRRRGGRRSRRPLR*GS
>read1/1|1R
PDPQRRGRRDLRPPRRRPPHRCQSLLGHRPRAR
>read1/1|2R
PTLNAEVAEICDHLAAARPTAVNLFWGIGRVRD
...
```

The UMGAP uses exact substring matching to identify reads. The next
step of the pipeline maps each 9-mer encountered in each of the protein
fragments via a provided index file. The available index files pair
each peptide from the UniProtKB onto the lowest common ancestor of all
proteins it occurs in.

```shell
$ umgap prot2kmer2lca -o -k9 9-mer.index < predicted-genes.faa | tee found-kmers.fa
>read1/1|1
0 0 0 0 0 0 0 0 0 0 35725 2759 1 1 1 2 515393 201174 0 1 0 0 0 379 0
>read1/1|2
0 0 0 0 0 0 0 0 0 1185650 1185650 140110 0 0 0 0 0 0 0 0 0 36842 0 1 162425
>read1/1|3
0 0 0 0 0 0 0 0 0 0 0 1773 47855 4530 1 1 1 1 1 0 35708 0 0 0
>read1/1|1R
0 0 0 0 2235 0 0 0 0 40674 1 1 2759 0 0 0 0 0 0 0 0 0 0 0 88724
>read1/1|2R
1 940557 1 940615 940615 940615 940615 940615 0 0 0 0 56 2 1 1 1 1 2 2 940615 940615 940615 940615 940615
...
```

The `-o` flag requests unrecognized peptides to be included as taxon ID
0. This sets the stage for the optional filtering of the taxa based on
their location: the seed extend tool.

```shell
$ umgap seedextend < found-kmers.fa | tee selected-seeds.fa
>read1/1|1
35725 2759 1 1 1 2 515393 201174
>read1/1|2
1185650 1185650 140110
>read1/1|3
1773 47855 4530 1 1 1 1 1
>read1/1|1R
40674 1 1 2759
>read1/1|2R
1 940557 1 940615 940615 940615 940615 940615 56 2 1 1 1 1 2 2 940615 940615 940615 940615 940615
...
```

These lists of taxa need to be aggregated into a single consensus
taxon per read. The `uniq` tool will gather all consecutive reads with
the same header after dropping a possible suffix, which is why the
`fastq2fasta` and `translate` steps must keep together the ends and
translations of a read. Next, the `taxa2agg` command aggregates the
complete set using a given taxonomy.

```shell
$ umgap uniq -d / < selected-seeds.fa | umgap taxa2agg taxons.tsv | tee classification.fa
>read1
940615
>read2
392734
>read3
332163
```

This output is in essence the result of the pipeline (and the same
result type of the preconfigured `umgap-analyse.sh` script). The
`umgap-visualize.sh` script internally uses two more commands: the
`taxa2freq` command to generate frequency tables and the `taxa2tree`
command to generate interactive visualizations.

```shell
$ umgap taxa2freq -r species taxons.tsv < classification.fa | tee report.csv
taxon id,taxon name,stdin
332163,Candidatus Solibacter usitatus,1
392734,Terriglobus roseus,1
940615,Granulicella tundricola,1
$ umgap taxa2tree --url < classification.fa # uses more reads for the image
https://bl.ocks.org/5960ffd859fb17439d7975896f513bc3
```

![A screenshot of the visualization created by the `taxa2tree` command.](treeview-advanced.png)

Running these commands in discrete steps, as above, is of course an
option, but the UMGAP wouldn't be a pipeline if it were the only option.
By putting all steps in a single command, they can run in parallel.

```shell
$ umgap fastq2fasta A1.fq A2.fq |          # joining paired-end files
  umgap translate -a |                     # translating all six frames
  umgap prot2kmer2lca -o -k9 9-mer.index | # mapping each 9-mer onto a taxon or 0
  umgap seedextend |                       # filtering extended seeds
  umgap uniq -d / |                        # joining equal headers (all taxa of the same read)
  umgap taxa2agg taxons.tsv |              # aggregating all taxa
  grep -v '^>' |                           # dropping headers
  umgap taxa2freq taxons.tsv               # make frequency table
taxon id,taxon name,stdin
332163,Candidatus Solibacter usitatus,1
392734,Terriglobus roseus,1
940615,Granulicella tundricola,1
```

### A comparative analysis {#section:comparative}

For a comparative study of 8 shotgun metagenomics samples and the
comparison of the taxonomic diversity therein [@detender], a comparative
frequency table was added to the UMGAP. Assuming the samples are stored
in pairs of files, following command will analyze all 8 paired-end
samples. By putting them together in a single analysis, the UMGAP can
reuse a memory-loaded index file.

```shell
$ umgap-analyse.sh \
  -1 ERR1654119_1.fastq.gz -2 ERR1654119_2.fastq.gz -t high-sensitivity -z -o ERR1654119.tid.fa.gz \
  -1 ERR1654120_1.fastq.gz -2 ERR1654120_2.fastq.gz -t high-sensitivity -z -o ERR1654120.tid.fa.gz \
  -1 ERR1654121_1.fastq.gz -2 ERR1654121_2.fastq.gz -t high-sensitivity -z -o ERR1654121.tid.fa.gz \
  -1 ERR1654122_1.fastq.gz -2 ERR1654122_2.fastq.gz -t high-sensitivity -z -o ERR1654122.tid.fa.gz \
  -1 ERR1654123_1.fastq.gz -2 ERR1654123_2.fastq.gz -t high-sensitivity -z -o ERR1654123.tid.fa.gz \
  -1 ERR1654124_1.fastq.gz -2 ERR1654124_2.fastq.gz -t high-sensitivity -z -o ERR1654124.tid.fa.gz \
  -1 ERR1654125_1.fastq.gz -2 ERR1654125_2.fastq.gz -t high-sensitivity -z -o ERR1654125.tid.fa.gz \
  -1 ERR1654126_1.fastq.gz -2 ERR1654126_2.fastq.gz -t high-sensitivity -z -o ERR1654126.tid.fa.gz
```

The usual individual visualizations are of course available and allow
studying each sample in detail. All 8 can be created using again a
single command, resulting in a URL for each sample.

```shell
$ umgap-visualize.sh -u *.tid.fa.gz
ERR1654119.tid.fa.gz: https://bl.ocks.org/5055456c85e9e103da221fbba6e6c269
ERR1654120.tid.fa.gz: https://bl.ocks.org/31e0fc6e2d8705e5ef64e856318dc17c
ERR1654121.tid.fa.gz: https://bl.ocks.org/019b89c96446a2949423d34f4281ee1c
ERR1654122.tid.fa.gz: https://bl.ocks.org/3542a2a24fcf5ed3a791e937d2d6e618
ERR1654123.tid.fa.gz: https://bl.ocks.org/ccc88f795dbd797b08cfe7c8d1187a13
ERR1654124.tid.fa.gz: https://bl.ocks.org/794f9907fbb2488d94f94ccb7d06d835
ERR1654125.tid.fa.gz: https://bl.ocks.org/7568952f1338768a6b62d32928c0ee74
ERR1654126.tid.fa.gz: https://bl.ocks.org/f25941840ce1e706623750d561d7c47d
```

Of more interest here is a comparative visualization. While the UMGAP
does not support such visualizations directly, it does support exporting
a comparative CSV frequency table.

```shell
$ umgap-visualize.sh -r phylum -t *.tid.fa.gz > frequencies.csv
$ head -10 frequencies.csv
taxon id,taxon name,ERR1654119.tid.fasta.gz,ERR1654120.tid.fasta.gz,ERR1654121.tid.fasta.gz,ERR1654122.tid.fasta.gz,ERR1654123.tid.fasta.gz,ERR1654124.tid.fasta.gz,ERR1654125.tid.fasta.gz,ERR1654126.tid.fasta.gz
1239,Firmicutes,83406,150700,80856,174901,80888,78400,95509,132121
2497438,Perkinsozoa,161,284,122,285,148,135,171,267
1752735,Candidatus Wolfebacteria,113,209,108,231,124,82,109,162
1817902,Candidatus Brennerbacteria,4,5,4,5,2,4,6,1
221216,Candidatus Parcubacteria,734,1485,823,1646,914,935,1304,1788
1817898,Candidatus Blackburnbacteria,28,66,26,57,18,41,49,68
3041,Chlorophyta,6439,11167,5921,13904,6303,6138,6897,10420
1913638,Zoopagomycota,1261,1730,937,2292,9727,6500,6728,7377
2732406,Kitrinoviricota,38,76,59,102,43,37,60,86
```

This CSV file can be processed by any spreadsheet software
or statistical analysis framework. For instance, in Figure
\ref{fig:comparative}, the fraction each of the 20 most detected
phyla represents in its sample is shown. For phyla uninfluenced by
the addition of chitin to the sample, we expect to see an equivalent
representation in each of the 8 samples (each color would take up 1/8th
of the bar). The Mucoromycota and Ciliophora phyla, however, are clearly
overrepresented in the 4 chitin-rich samples.

![The relative representation of the 20 most frequent phyla (over all samples) per sample. Phyla with equivalent representation in all 8 samples result in an approximately uniform division of the bar over the samples. Notable are the Ciliophora and Mucoromycota phyla, which are overrepresented in the 4 latter samples, which are the chitin-rich samples](./comparative-phyla.svg){#fig:comparative}

### A transcriptomics analysis pipeline {#section:transcript}

By request of Steve Baeyen, the script below was written to analyse a
number of transcriptomics samples (stored in `samples`). This script is
very similar to a specialized `umgap-analyse` script without the various
parameters and configurations. However, since these are transcriptomics
samples, instead of joining all reads after identification, the `umgap
bestof` tool selects the single best frame out of the 6 translated and
identified frames.

```sh
#!/bin/sh
samples="LJ_2017_035 Rhizo_P1 ..."
ninemers=/.../ninemer.fst
taxons=/.../taxons.tsv

d="$(mktemp -d)"

umgap prot2kmer2lca -m -o -s "$d/socket" "$ninemers" &
ninemer_pid="$!"
trap "rm -rf '$d' && kill '$ninemer_pid'" KILL EXIT
while [ ! -S "$d/socket" ] && sleep 1; do true; done

mkdir -p results

for sample in $samples; do
  mkfifo "$d/1" "$d/2"
  zcat "data/${sample}_1.fastq.gz" > "$d/1" &
  zcat "data/${sample}_2.fastq.gz" > "$d/2" &
  umgap fastq2fasta "$d/1" "$d/2"   | # intercalating ends
    umgap translate -n -a -t11      | # translating 6 frames
    nc -NU "$d/socket"              | # identification
    umgap bestof                    | # pick best frame
    umgap uniq -d ' '               | # drop end and frame indicator
    umgap taxa2agg "$taxons"        | # aggregate
    gzip - > results/"${sample}.gz"
  rm "$d/1" "$d/2"
done
```

This scripts outputs a single file per sample. These files can then
be processed with the visualize script into a frequency table. This
CSV table contains a row per predicted taxon, and columns with the
taxon identifier, name and for each of the files, the number of reads
attributed to this taxon in the corresponding sample.

```shell
$ umgap-visualize -t results/*
taxon id,taxon name,LJ_2017_033.gz,LJ_2017_035.gz,LJ_2017_036.gz,LJ_2017_039.gz,LJ_2017_040.gz,PS1_P1.gz,PS1_P2.gz,Rhizo_P1.gz,Rhizo_P2.gz
1,root,18317594,16378552,16425896,17262443,17494244,17475641,13965489,23492453,16903317
562,Escherichia coli,641334,557683,92129,373881,168358,279683,226576,237624,171574
195883,Laodelphax striatellus,15227,11485,1832336,14107,18741,12177,8955,14312,14004
1913989,Gammaproteobacteria bacterium,159917,142469,97749,162877,165087,400512,140581,265984,289616
1978231,Acidobacteria bacterium,350326,252945,106724,397893,301891,58982,47283,89638,51127
7070,Tribolium castaneum,2759,2437,1481795,4262,3392,5893,5150,9191,5110
2026735,Deltaproteobacteria bacterium,140640,123579,56559,160855,149663,165955,132033,201829,138533
2026724,Chloroflexi bacterium,144053,136295,111208,172663,144990,103864,67742,150339,74256
1883427,Actinobacteria bacterium,135642,126913,54918,164617,152087,108761,126096,133689,101171
3870,Lupinus albus,244425,195462,75007,146726,148291,42949,63598,85444,95046
...
```

<!--
```
	bacteriophage	mycrovirus	invertebrates viruses	plant_virus
LJ_2017_033	Sulfolobales Virus YNP2 strain SYV2	(Soybean leaf-associated mitovirus 5 isolate SaMitV5-1)	Beihai levi-like virus, Hubei levi-like virus	[Beet cryptic virus 2]
LJ_2017_033_No_Dnase	Sulfolobales Virus YNP2 strain SYV2	Soybean leaf-associated mitovirus 5 isolate SaMitV5-1	Wenling levi-like virus 1 strain	[Beet cryptic virus 2]
LJ_2017_034	Sulfolobales Virus YNP2 strain SYV2	Phakopsora pachyrhizi mycovirus isolate,	Beihai levi-like virus 10 strain	[Beet cryptic virus 2]
LJ_2017_034_No_Dnase	Sulfolobales Virus YNP2 strain SYV2	Niflavirus isolate NiflaV	Beihai levi-like virus 10 strain	(Raphanus sativas chrysovirus 1)
LJ_2017_035	Sulfolobales Virus YNP2 strain SYV2	Plasmopara halstedii virus A	Beihai levi-like virus 7 strain	[Beet cryptic virus 2]
LJ_2017_036	Sulfolobales Virus YNP2 strain SYV2	Macrophomina phaseolina single-stranded RNA virus 1 isolate Tn408	Beihai levi-like virus 11	[Beet cryptic virus 2]
LJ_2017_039	Sulfolobales Virus YNP2 strain SYV2	Plasmopara halstedii virus A isolate Ph8-99	Hubei noda-like virus	[Beet cryptic virus 2]
LJ_2017_040	Sulfolobales Virus YNP2 strain SYV2	Plasmopara halstedii virus A isolate Ph8-99	Beihai levi-like virus	[Beet cryptic virus 2]
PS1_P1	Pyrobaculum filamentous virus 1	Niflavirus isolate NiflaV	Wenzhou levi-like virus 2	Lettuce mottle virus isolate
PS1_P2	Sulfolobales Virus YNP2 strain SYV2	Plasmopara halstedii virus A isolate Ph8-99	Sanxia noda-like virus 1	Lettuce mottle virus isolate
Rhizo_P1	Sulfolobales Virus YNP2 strain SYV2	Niflavirus isolate NiflaV	Changjiang picorna-like virus	Soybean-associated bicistronic virus
Rhizo_P2	Pyrobaculum filamentous virus 1	Niflavirus isolate NiflaV, Diaporthe ambigua RNA virus 1	Beihai levi-like virus 3	(Soybean-associated bicistronic virus)

strain	lage similariteiten
strain	(betrouwbaar)
strain	[goede detectie]
```
-->

<!--
Voor het ILVO (Steve Baeyen & Caroline):

```
/data/mail/ugent/Inbox/cur/1526652704.11715_1.abysm,U=11303:2,RS
/data/mail/ugent/Sent/cur/1527859599.M299129P1477Q1.abysm,U=775:2,S
/data/mail/ugent/Sent/cur/1529326023.M513732P13462Q1.abysm,U=819:2,S
/data/mail/ugent/Inbox/cur/1533069604.1482_25.abysm,U=12046:2,RS
/data/mail/ugent/Sent/cur/1534162605.M622864P14331Q1.abysm,U=885:2,S
```

Voor het ITG:

```
/data/mail/gmail/Inbox/cur/1494574491.1508_9.abysm,U=17583:2,S
/data/mail/gmail/Sent/cur/1494601367.M458592P22678Q1.abysm,U=3002:2,S
```

Voor het ILVO (Annelies Haegeman):

```
/data/mail/ugent/Inbox/cur/1562229558.23101_1.abysm,U=19052:2,RS
/data/mail/ugent/Inbox/cur/1562322121.3910_1.abysm,U=19074:2,RS
/data/mail/ugent/Sent/cur/1564754751.M820714P3686Q1.abysm,U=1610:2,S
/data/mail/ugent/Inbox/cur/1565085863.24367_2.abysm,U=19459:2,RS
/data/mail/ugent/Sent/cur/1565166804.M449717P11530Q1.abysm,U=1623:2,S
/data/mail/ugent/Inbox/cur/1565190743.13909_1.abysm,U=19575:2,S
```
-->

### Finding viral traces {#section:viral}

TODO drop?

<!--
Sebastiaan Theuns https://biblio.ugent.be/person/0767592A-F0EE-11E1-A9DE-61C894A0A6B4

```
/data/mail/ugent/Inbox/cur/1559131238.16965_1.abysm,U=18505:2,RS
```
-->
