## Case studies and analyses {#section:casestudies}

During development, the still incomplete pipeline was applied to several
datasets as a means of exploration. This has helped to find bugs,
usability problems, discover other applications and guide the UMGAP to
its final form. This section describes the analysis of these datasets,
as it would be done using the released version of the UMGAP, not as it
was executed originally. As such, this section may be used as a guide to
compose a new pipeline for a similar analysis.

* TODO Annelies Haegeman https://biblio.ugent.be/person/F7927142-F0ED-11E1-A9DE-61C894A0A6B4

### The preconfigured pipelines {#section:preconf}

The primary and easiest procedure to run the UMGAP on your data, and the
last addition to the pipeline, is by using one of the six preconfigured
pipelines. They are six diverging ways of configuring the pipeline,
picked to cover the most probable use cases. Should your data fall within
such a use case, running the pipeline is reduced to running two or three
scripts.

The first is a setup script called `umgap-setup.sh`. It is an
interactive script which helps you get the required databases and
optional external tools in the right place. In general, the setup script
only needs to run once, but it can also be used to verify the file
locations and download newer versions of the data.

Following code snippet shows the interaction with the setup script.
It starts out by asking the relevant questions, without delay. At the
end, it downloads the relevant files in sequence, without further
interaction. In this snippet, we choose to download the taxonomy and the
tryptic index, as this example will use a tryptic pipeline.

```shell
$ umgap-setup.sh -f /opt/FragGeneScanPlusPlus
Use '/home/user/.config/unipept' as configuration directory? [y]/n y
Created directory /home/user/.config/unipept.
Found, tested and remembered the FragGeneScan++ location.
Use '/home/user/.local/share/unipept' as data directory? [y]/n y
Created directory /home/user/.local/share/unipept.
Checking the latest version on the server.
Latest version is 2020-12-02.

For any type of analysis, you need a taxonomony. For mapping tryptic or
9-mer peptides to it, you need the respective index file of the same
version.

Would you like to download the taxonomy from 2020-12-02 (115MiB)? [y]/n y
Would you like to download the tryptic index from 2020-12-02 (12GiB)? [y]/n y
Would you like to download the 9-mer index from 2020-12-02 (152GiB)? [y]/n n
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  114M  100  114M    0     0  6477k      0  0:00:18  0:00:18 --:--:-- 6332k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   12G  100   12G    0     0  5076k      0  0:34:12  0:34:12 --:--:-- 5790k
```

The script also allows further configuration via some flags.

`-c dir`
  ~ Set a different configuration directory.

`-f dir`
  ~ Link the location of your FragGeneScan installation.

`-d dir`
  ~ Set a different location for downloading the database files. This
    can be on a separate disk.

`-y`
  ~ Download all files of the latest version without asking for
    confirmation.

This last flag, `-y`, is especially useful when running the setup script
from another script, e.g. an automatic update script.

The second script is the main analysis script, aptly called
`umgap-analyse.sh`. It will run a selected pipeline on the input data
using the databases downloaded by the setup script. Consider for
instance a 100 paired-end reads sampled from a dataset generated by
@lindgreen for their MetaBenchmark. With these paired-end reads in the
`A1.fq` and `A2.fq` files, the commands are as follows.

```shell
$ head -4 A1.fq
@1198114###CP002480-_Acidobacteria_733918/1
ATCGCGCACGCGGCCGATGCCCCAGAAGAGATTGACAGCGGTGGGGCGGGCGGCGGCGAGGTGGTCGCAGATCTCGGCGACCTCTGCGTTGAGGGTCGGG
+
AADEGGAGIIEIHFHKHJKKJKHIJJIIKCJAHBFKKFHIJIF;JIH5DE$I>CBE$E:FGEJCGABEEECCDKD?D?C$ECEEE;CEEDEEAD=?ED$D
$ umgap-analyse.sh -1 A1.fq -2 A2.fq -t tryptic-precision -z -o tryptic-precision-output.fa.gz
$ zcat tryptic-precision-output.fa.gz | head -2
>1198114###CP002480-_Acidobacteria_733918/1_1_100_-
1
```

It seems the pipeline cannot be more specific than root (Taxon ID 1) for
the first read.

The analysis script was passed a single series of arguments here (`-1`
to through `-o`), but it accepts multiple such sequences (along with the
other flags specified below) to combine multiple analyses into a single
command.

`-c dir`
  ~ Set a different configuration directory.

`-1 file`
  ~ Single ended FASTA or first pair-ended FASTQ input file, optionally
    compressed.

`-2 file`
  ~ Second pair-ended FASTQ input file, optionally compressed.

`-t`
  ~ Type of the analysis. This flag allows you to select one of the preconfigured pipelines.

    - `high-precision` is the default pipeline and is optimized for high
      precision identifications on your metagenomics reads.

    - `max-precision` is optimized for highest precision, with a small
      setback on sensitivity.

    - `tryptic-precision` is made for fast analyses on your laptop.
      Fewer results, but accurate.

    - `high-sensitivity` is optimized for high sensitivity
      identifications on metagenomics reads.

    - `max-sensitivity` is optimized for highest sensitivity, with a
      small setback on precision.

    - `tryptic-sensitivity` is made for fast analyses on your laptop.
      Get a quick overview of your data.

`-z`
  ~ Request compression of the output file.

`-o file`
  ~ The output file.

The output file contains a single taxonomic prediction for every read
after the analysis script. To interpret these results without parsing
and scripting visualizations, the `umgap-visualize.sh` script is
provided. This visualization script can create importable CSV frequency
tables and interactive visualizations. The latter can be stored locally
or hosted online. The following snippet creates all three in turn.

```shell
$ umgap-visualize.sh -t -r phylum tryptic-precision-output.fa.gz
taxon id,taxon name,tryptic-precision-output.fa.gz
1117,Cyanobacteria,2
57723,Acidobacteria,4
201174,Actinobacteria,4
1224,Proteobacteria,5
1,root,160
$ umgap-visualize.sh -w tryptic-sensitivity-output.fa.gz > tryptic-sensitivity.html
$ umgap-visualize.sh -u tryptic-sensitivity-output.fa.gz
tryptic-sensitivity-output.fa.gz: https://bl.ocks.org/11b7809d6754b9530cf1a49d93a8d568
```

The CSV tables contain a record for each taxon found in the sample.
A record contains, in order, the taxon ID, (for convenience) the
taxon name and the number of reads assigned to this taxon or below.
All records will be at the same specified taxon rank or at root
(unidentified).

`-t`
  ~ Output a CSV frequency table on species rank.
`-w`
  ~ Output an HTML web page of an interactive visualization.
`-u`
  ~ Print a shareable URL to a online interactive visualization.
`-r rank`
  ~ Set the rank for the CSV frequency table (default: species).

### A home brew taxonomic analysis pipeline {#section:homebrew}

While the preconfigured pipelines make running the UMGAP relatively
easy, there are only six of them. Part of the reason for using a
pipeline is to allow composing many diverse metagenomic pipelines
with the same set of tools. This section builds up a custom pipeline,
demonstrating how all tools work together.

The 3 reads of the same dataset as in \ref{section:preconf} will
be used. This is a paired-end dataset, stored in two FASTQ files.
Since the pipeline operates on FASTA files, the first step is a format
conversion.

```shell
$ umgap fastq2fasta A1.fq A2.fq | tee preprocessed.fa
>read1/1
ATCGCGCACGCGGCCGATGCCCCAGAAGAGATTGACAGCGGTGGGGCGGGCGGCGGCGAGGTGGTCGCAGATCTCGGCGACCTCTGCGTTGAGGGTCGGG
>read1/2
AAGATGGCGACTGGATGATGATCGCGCGGCAGGCCACCATCTACGATCCCGCAGTGAAGCATTACTAACCATGATGCGCAACGACTCGCTGTCAGAGCTA
>read2/1
AGATTGCTGGTGCGGGTGCTCTGCCGGGCTTCTTCATCTCGGACCGCGCATCCGATGCGCACACGGCACAGTAGGTATGCGGTGAGAGCACCTCGCTTTT
>read2/2
CGCAATCTTGCGGCGCACCGCGATCAAATCGGGATAGTCCGGTGTGTAACGCGTGCTCAGGTCATCTGCCCGTACCCGCAATACATTCAACTGTGTTTTA
>read3/1
CCCTCAAGCCCGATCGAACTTTCCTACTGCCCGACCTCAGCGCGGAGCATTACCGAATCGTCGTCAACAATCTCCCCGATGGCTTCTATGTGAACTCCAT
>read3/2
CAAAATGGAGTGGGCGCTACTGCTCCGTGAGCAGGGTCAGTTGGAGGGATTGGGGCGTGCGCTCGGGGATGCTAATGGCGGTGCCCCTGGATTCCTGAGA
```

This command interleaves the given FASTQ files to output a single
FASTA stream, in which the paired ends alternate. This FASTA stream is
a stream of DNA reads. Each read needs to be translated to an amino
acid sequence. While any gene prediction tool could be used here, the
UMGAP was tested with FragGeneScan and its variants.

```shell
$ FGS -s preprocessed.fa -o predicted-genes -w0 -t illumina_10 -p 16 > /dev/null
$ rm predicted-genes.fnn predicted-genes.out # we don't use these files
```

Alternatively, the built-in translation tool can be used to translate
all six frames of each read. Whichever method used, it results in a file
containing protein fragments.

```shell
$ umgap translate -a < preprocessed.fa | tee predicted-genes.faa
>read1/1|1
IAHAADAPEEIDSGGAGGGEVVADLGDLCVEGR
>read1/1|2
SRTRPMPQKRLTAVGRAAARWSQISATSALRVG
>read1/1|3
RARGRCPRRD*QRWGGRRRGGRRSRRPLR*GS
>read1/1|1R
PDPQRRGRRDLRPPRRRPPHRCQSLLGHRPRAR
>read1/1|2R
PTLNAEVAEICDHLAAARPTAVNLFWGIGRVRD
...
```

The UMGAP uses exact substring matching to identify reads. The next
step of the pipeline maps each 9-mer encountered in each of the protein
fragments via a provided index file. The available index files pair
each peptide from the UniProtKB onto the lowest common ancestor of all
proteins it occurs in.

```shell
$ umgap prot2kmer2lca -o -k9 9-mer.index < predicted-genes.faa | tee found-kmers.fa
>read1/1|1
0 0 0 0 0 0 0 0 0 0 35725 2759 1 1 1 2 515393 201174 0 1 0 0 0 379 0
>read1/1|2
0 0 0 0 0 0 0 0 0 1185650 1185650 140110 0 0 0 0 0 0 0 0 0 36842 0 1 162425
>read1/1|3
0 0 0 0 0 0 0 0 0 0 0 1773 47855 4530 1 1 1 1 1 0 35708 0 0 0
>read1/1|1R
0 0 0 0 2235 0 0 0 0 40674 1 1 2759 0 0 0 0 0 0 0 0 0 0 0 88724
>read1/1|2R
1 940557 1 940615 940615 940615 940615 940615 0 0 0 0 56 2 1 1 1 1 2 2 940615 940615 940615 940615 940615
...
```

The `-o` flag requests unrecognized peptides to be included as taxon ID
0. This sets the stage for the optional filtering of the taxa based on
their location: the seed extend tool.

```shell
$ umgap seedextend < found-kmers.fa | tee selected-seeds.fa
>read1/1|1
35725 2759 1 1 1 2 515393 201174
>read1/1|2
1185650 1185650 140110
>read1/1|3
1773 47855 4530 1 1 1 1 1
>read1/1|1R
40674 1 1 2759
>read1/1|2R
1 940557 1 940615 940615 940615 940615 940615 56 2 1 1 1 1 2 2 940615 940615 940615 940615 940615
...
```

These lists of taxa need to be aggregated into a single consensus
taxon per read. The `uniq` tool will gather all consecutive reads with
the same header after dropping a possible suffix, which is why the
`fastq2fasta` and `translate` steps must keep together the ends and
translations of a read. Next, the `taxa2agg` command aggregates the
complete set using a given taxonomy.

```shell
$ umgap uniq -d / < selected-seeds.fa | umgap taxa2agg taxons.tsv | tee classification.fa
>read1
940615
>read2
392734
>read3
332163
```

This output is in essence the result of the pipeline (and the same
result type of the preconfigured `umgap-analyse.sh` script). The
`umgap-visualize.sh` script internally uses two more commands: the
`taxa2freq` command to generate frequency tables and the `taxa2tree`
command to generate interactive visualizations.

```shell
$ umgap taxa2freq -r species taxons.tsv < classification.fa | tee report.csv
taxon id,taxon name,stdin
332163,Candidatus Solibacter usitatus,1
392734,Terriglobus roseus,1
940615,Granulicella tundricola,1
$ umgap taxa2tree --url < classification.fa # uses more reads for the image
https://bl.ocks.org/5960ffd859fb17439d7975896f513bc3
```

![A screenshot of the visualization created by the `taxa2tree` command.](treeview-advanced.png)

Running these commands in discrete steps, as above, is of course an
option, but the UMGAP wouldn't be a pipeline if it were the only option.
By putting all steps in a single command, they can run in parallel.

```shell
$ umgap fastq2fasta A1.fq A2.fq |          # joining paired-end files
  umgap translate -a |                     # translating all six frames
  umgap prot2kmer2lca -o -k9 9-mer.index | # mapping each 9-mer onto a taxon or 0
  umgap seedextend |                       # filtering extended seeds
  sed '/^>/s_/.*__' |                      # dropping the paired-end and frame annotations
  umgap uniq |                             # joining equal headers (all taxa of the same read)
  umgap taxa2agg taxons.tsv |              # aggregating all taxa
  grep -v '^>' |                           # dropping headers
  umgap taxa2freq taxons.tsv               # make frequency table
taxon id,taxon name,stdin
332163,Candidatus Solibacter usitatus,1
392734,Terriglobus roseus,1
940615,Granulicella tundricola,1
```

### A comparative analysis {#section:comparative}

Peat substrate amended with chitin modulates the N-cycle, siderophore and chitinase responses in the lettuce rhizobiome

### A transcriptomics analysis pipeline {#section:transcript}

Steve Baeyen https://biblio.ugent.be/publication?q=author%3D%22Baeyen%2C+Steve%22

### Finding viral traces {#section:viral}

Sebastiaan Theuns https://biblio.ugent.be/person/0767592A-F0EE-11E1-A9DE-61C894A0A6B4
